src/lib.rs

```diff
@@ -358,22 +385,30 @@ impl CANSocket {
     /// Blocking read a single can frame.
     pub fn read_frame(&self) -> io::Result<CANFrame> {
         let mut frame = CANFrame {
-            _id: 0,
-            _data_len: 0,
-            _pad: 0,
-            _res0: 0,
-            _res1: 0,
-            _data: [0; 8],
+            tag: CANFrameType::Normal,
+
+            /// NOTE: Can't use `std::mem::MaybeUninit::uninit().assume_init()`, which
+            /// would be very appropriate here, because that would result in undefined
+            /// behavior, according to `std::mem::MaybeUninit` docs (1.36.0):
+            ///
+            /// > [...] uninitialized memory is special in that the compiler knows that it
+            /// > does not have a fixed value. This makes it undefined behavior to have
+            /// > uninitialized data in a variable even if that variable has an integer type,
+            /// > which otherwise can hold any fixed bit pattern [...]
+            ///
+            s: unsafe { std::mem::MaybeUninit::zeroed().assume_init() }
```

Comment:
I think the correct way to use ```MaybeUninit``` in this case would be as following: 
1. ```let mut frame: MaybeUninit<CANFrameStruct> = MaybeUninit::uninit();```
2. Perform the syscall. 
3. If return value is okay, check if FD or not (store in ```tag```) and then return: 
```
Ok(CANFrame {
    tag, 
    s: unsafe { frame.assume_init() }  
})
```

Reply:
Well, I wish that would work... But is there really guarantee that's not UB?

Reply:
What are your specific concerns? To me, there are 3 potential sources of UB, and I will briefly discuss why I do not think that they apply: 

1. Creating a value with an illegal byte layout is UB (example in the documentation https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html is a reference, which may never be entirely null on the byte level). This is not relevant for our type, as all contained types accept all possible byte layouts (they are primitives that are not bool). 
2. Using ```MaybeUninit::assume_init()``` to avoid invariants of a type, for example established by a constructor, can lead to UB. This is not relevant, as this is a POD without any actual semantics and particularly, also without a ```Drop``` implementation, as it only a C wrapper. 
3. Creating a variable from memory that has not been initialized leads to its value not being "fixed" and interacting with such a variable is UB. Here, we rely on the guarantee of the sys call that the C code actually properly initializes all the fields (when the return value indicates success). Personally, I think that is a reasonable assumption, although I have to admit I am not a 100% sure about the padding bytes. However, one might argue that we are not actually reading from or writing to these values. It is fair to argue that looking only at the header, with the underlying implementation prone to change, we cannot actually guarantee this. In this case, we would have to go with an initialized version of the struct. 

As long as the type is wrapped in a ```MaybeUninit```, it is fine for the memory to not be initialized, i.e. violate the above requirements. So the only risk occurs on the ```assume_init()```. Btw, in C code, it is perfectly fine to write to uninitialized memory (reading from it is also UB in C). 

An important distinction to take not of is that I am explicitly not making the type of the variable be ```MaybeUninit<CANFrame>```, but ```MaybeUninit<CANFrameStruct>```. A type containing an Uninit type might have different layout than the same type with just the nested type in it. This, however, would not even compile here as ```CANFrame``` is not generic.

Reply:
My concern is about the compiler assumptions. I don't doubt, that the C function called via FFI actually do the right thing with the pointed memory. But does the Rust compiler really get to know that the memory got written and doesn't do some kind of reordering resulting in returning invalid values regardless the actual memory content?

From the examples in the documentation (https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html) they always do some kind of write operation before claiming `assume_init` on the memory. But can a FFI call be counted as such write?